src/main/java/tigase/server/amp/db/MsgRepository.java

```diff
@@ -288,22 +300,88 @@ public boolean belongsTo(Class<? extends BasicComponent> component) {
 
 		@Override
 		public Element getMessageExpired(long time, boolean delete) {
-			lock.lock();
-			try {
+			// what if some queue has entries one repo (far in the future)
+			// but the other repo loaded only part of his queue and some remained in the database?
+
+			if (expiredQueue.size() == 0) {
+
+				// If the queue is empty load it with some elements
+				loadExpiredQueue(MAX_QUEUE_SIZE);
+			} else {
+				// Check if any repository in the poll has empty expiredQueue and if so, try to load for them..
 				for (MsgRepositoryIfc repo : getRepositories().values()) {
-					Element el = repo.getMessageExpired(time, delete);
-					if (el != null) {
-						return el;
+					if (repo instanceof MsgRepository && ((MsgRepository) repo).awaitingInExpiredQueue.get() == 0) {
+						((MsgRepository) repo).loadExpiredQueue(MAX_QUEUE_SIZE);
 					}
 				}
-				expiredMessagesCondition.await();
-			} catch (InterruptedException e) {
-				log.log(Level.FINER, "awaiting for expired messages interrupted");
-			} finally {
-				lock.unlock();
+
+				// If the queue is not empty, check whether recently saved off-line
+				// message
+				// is due to expire sooner then the head of the queue.
+				RepoAwareMsgDBItem item = expiredQueue.peek();
+
+				if ((item != null) && (earliestOffline < item.expired.getTime())) {
+
+					// There is in fact off-line message due to expire sooner then the head
+					// of the
+					// queue. Load all off-line message due to expire sooner then the first
+					// element
+					// in the queue.
+					loadExpiredQueue(item.expired);
+				}
 			}
 
-			return null;
+			RepoAwareMsgDBItem item = null;
+			while (item == null) {
+				try {
+					item = expiredQueue.take();
+				} catch (InterruptedException ex) {
+				}
+			}
+
+			if (item != null && item.getRepo() instanceof MsgRepository) {
```

Comment:
On the one hand - we shouldn't get null item object here, on the other if we are already checking it - shouldn't we move the below condition (`if(delete`) within this if condition?

Reply:
I've removed this check as it was pointless.